packages/playwright-core/src/server/screenshotter.ts

```diff
@@ -107,50 +116,92 @@ export class Screenshotter {
-    progress.throwIfAborted(); // Screenshotting is expensive - avoid extra work.
-    const shouldSetDefaultBackground = options.omitBackground && format === 'png';
-    if (shouldSetDefaultBackground) {
-      await this._page._delegate.setBackgroundColor({ r: 0, g: 0, b: 0, a: 0 });
-      progress.cleanupWhenAborted(() => this._page._delegate.setBackgroundColor());
-    }
-    progress.throwIfAborted(); // Avoid extra work.
-
-    const restoreBlinkingCaret = async () => {
-      await Promise.all(this._page.frames().map(async frame => {
-        frame.nonStallingEvaluateInExistingContext('window.__cleanupScreenshot && window.__cleanupScreenshot()', false, 'utility').catch(() => {});
-      }));
-    };
+  async _preparePageForScreenshot(progress: Progress, disableAnimations: boolean) {
     await Promise.all(this._page.frames().map(async frame => {
-      await frame.nonStallingEvaluateInExistingContext((function() {
+      await frame.nonStallingEvaluateInExistingContext('(' + (function(disableAnimations: boolean) {
         const styleTag = document.createElement('style');
         styleTag.textContent = `
-          * { caret-color: transparent !important; }
-          * > * { caret-color: transparent !important; }
-          * > * > * { caret-color: transparent !important; }
-          * > * > * > * { caret-color: transparent !important; }
+          *,
+          * > *,
+          * > * > *,
+          * > * > * > *,
           * > * > * > * > * { caret-color: transparent !important; }
         `;
         document.documentElement.append(styleTag);
+        const infiniteAnimationsToResume: Animation[] = [];
+        const cleanupCallbacks: (() => void)[] = [];
+
+        if (disableAnimations) {
+          const collectRoots = (root: Document | ShadowRoot, roots: (Document|ShadowRoot)[] = []): (Document|ShadowRoot)[] => {
+            roots.push(root);
+            const walker = document.createTreeWalker(root, NodeFilter.SHOW_ELEMENT);
+            do {
+              const node = walker.currentNode;
+              const shadowRoot = node instanceof Element ? node.shadowRoot : null;
+              if (shadowRoot)
+                collectRoots(shadowRoot, roots);
+            } while (walker.nextNode());
+            return roots;
+          };
+          const handleAnimations = (root: Document|ShadowRoot): void => {
+            for (const animation of root.getAnimations()) {
+              try {
+                animation.finish();
+              } catch (e) {
+                animation.cancel();
+                infiniteAnimationsToResume.push(animation);
+              }
+            }
+          };
+          for (const root of collectRoots(document)) {
+            const handleRootAnimations: (() => void) = handleAnimations.bind(null, root);
+            handleRootAnimations();
+            root.addEventListener('transitionrun', handleRootAnimations);
+            root.addEventListener('animationstart', handleRootAnimations);
```

Comment:
Won't this go over all animations once again and put them into `infiniteAnimationsToResume` list multiple times?

Reply:
cancelling infinite animation synchronously remove it from `document.getAnimations()` list. However, we can use `Set` here instead to be on the safe side!